### PR TITLE
refactor(runtime): pass initial root element directly

### PR DESCRIPTION
**Pull request by arvinpoddar**
Rather than passing metadata about whether the root element exists or not, we now allow the root element to be null, and also pass an initial root element field. This allows us to derive whether the root exists, as well as use this initial root to reset the data later.

### CODE REVIEW

File: packages/runtime/src/state/actions.ts [hunk 1]
```diff
@@ -99,10 +99,15 @@ type DocumentPayloadBaseDocument = {
...
+  key: string
+  locale?: string | null // older versions of the runtime may not provide this field
   id: string
   type: string
   name: string
+  rootElement: Element | null
+  initialRootElement: Element
+  hasFallback?: boolean
   __type: typeof EMBEDDED_DOCUMENT_TYPE
 }
```
Comment (arvinpoddar) [inline packages/runtime/src/state/actions.ts:112]: This is a breaking change in the data type being transacted between the host and builder. Is it permissible to make this change since embedded components are not being used?

Comment (fikrikarim) [reply packages/runtime/src/state/actions.ts:112]: Not used by users? AFAIK, people are using it for the one-click catalyst demo.

Comment (arvinpoddar) [reply packages/runtime/src/state/actions.ts:112]: Ah, let me rephrase: "not being used in general acceptable". Since the OCC demo users are limited, is it acceptable to tell them to upgrade their runtime instead of having to support backwards compatibility?

Comment (fikrikarim) [reply packages/runtime/src/state/actions.ts:112]: I'm not sure about that. Let's talk about this during stand-up.

**Review by fikrikarim — DISMISSED** [review]
LGTM if we're okay with the breaking change

File: packages/runtime/src/state/actions.ts (hunk 1 above)
Comment (migueloller) [reply packages/runtime/src/state/actions.ts]: What did you conclude after chatting with stakeholders @arvinpoddar?

File: packages/runtime/src/next/client.ts [hunk 2]
```diff
@@ -175,35 +175,36 @@ export function componentDocumentToRootEmbeddedDocument({
...
 
-  const rootElement = data ?? {
+  const initialRootElement = {
     // Fallback rootElement
     // Create a stable uuid so two different clients will have the same empty element data.
     // This is needed to make presence feature work for an element that is not yet created.
-    key: deterministicUUID({ id, locale, seed: document.seed }),
+    key:
+      rootElement != null
+        ? rootElement?.key
```
Comment (migueloller) [inline packages/runtime/src/next/client.ts:192]: Optional chaining is unnecessary here since we check `rootElement` is not `null`

File: packages/runtime/src/runtimes/react/components/Document.tsx [hunk 3]
```diff
@@ -15,15 +15,14 @@ export const Document = memo(
     { document, fallback }: DocumentProps,
     ref: Ref<ElementImperativeHandle>,
   ): JSX.Element {
-    const isInitialData = match(document)
-      .with({ meta: { isInitialData: true } }, () => true)
-      .otherwise(() => false)
+    const isInitialData = document.rootElement == null
+    const rootElement = getDocumentRootElement(document)
```
Comment (migueloller) [inline packages/runtime/src/runtimes/react/components/Document.tsx:19]: Isn't `getDocumentRootElement` a selector? I don't think we should "pierce" into modules under the `state` directory as this is meant to be encapsulated by the interface the wraps the Redux store. Also, selectors should be reserved to be used with Redux state, not as general getters when we have data that matches the state. We only want to have to change selectors when the state changes, and not couple it to this part of the codebase, which isn't dealing with that state module

File: packages/runtime/src/state/actions.ts (hunk 1 above)
Comment (arvinpoddar) [reply packages/runtime/src/state/actions.ts]: @migueloller Chatted with Nathan, and the answer is **it should be possible** to force all OCC hosts to update. We'd have to do it with a script that crawls the existing projects and redeploy them. Given that there are nearly 200 OCC hosts that would be affected, I didn't feel great about this option - I would've been okay if it had been a few dozen. As much as I didn't want backwards compatibility to be a concern so early, I don't see another way.

File: packages/runtime/src/runtimes/react/components/Document.tsx (hunk 3 above)
Comment (migueloller) [reply packages/runtime/src/runtimes/react/components/Document.tsx:19]: I noticed later that this isn't a selector. The naming is unfortunate, though, since it follows the same naming convention there. We also seem to be using it similar to selectors elsewhere (`Documents.getRootElement`). I think we should place this in a shared place that is not the `state` directory so that we can keep things in those modules encapsulated.

**Review by migueloller — CHANGES_REQUESTED** [review]
> This allows us to derive whether the root exists, as well as use this initial root to reset the data later.

In what cases do we need to "reset the data later"?

File: packages/runtime/src/runtimes/react/components/Document.tsx (hunk 3 above)
Comment (arvinpoddar) [reply packages/runtime/src/runtimes/react/components/Document.tsx:19]: Fair point, will address.

**Comment by arvinpoddar** [comment]
> > This allows us to derive whether the root exists, as well as use this initial root to reset the data later.
> 
> In what cases do we need to "reset the data later"?

"Reset" refers to replacing the root element with the initial root element. This mechanic will likely be used for resetting to base locale and deletion (triggered by a message sent from the builder).

File: packages/runtime/src/next/client.ts (hunk 2 above)
Comment (arvinpoddar) [reply packages/runtime/src/next/client.ts:192]: Good catch, I was trying to do a type deduction with the optional chaining at first, must've left it behind.

**Comment by migueloller** [comment]
> > > This allows us to derive whether the root exists, as well as use this initial root to reset the data later.
> > 
> > 
> > In what cases do we need to "reset the data later"?
> 
> "Reset" refers to replacing the root element with the initial root element. This mechanic will likely be used for resetting to base locale and deletion (triggered by a message sent from the builder).

So in this case the initial document wouldn't be the empty element but instead the base locale's element? I thought we were going to use deletion of the element tree as reset to base so that things wouldn't stay "forked"

File: packages/runtime/src/runtimes/react/components/Document.tsx (hunk 3 above)
Comment (arvinpoddar) [reply packages/runtime/src/runtimes/react/components/Document.tsx]: Ended up making a separate directory within `state` to store this utility function. It felt weird to me to remove this from `state` when the function operates on a type that was defined in a `state` module - moreover, all usages (except for one) of this function are within `state` as well.

**Review by Codeseph — APPROVED** [review]

**Review by migueloller — APPROVED** [review]
